airflow/hooks/hdfs_hook.py

```diff
@@ -17,85 +17,131 @@
 
     def get_conn(self):
-        """
-        Returns a snakebite HDFSClient object.
-        """
-        # When using HAClient, proxy_user must be the same, so is ok to always
-        # take the first.
-        effective_user = self.proxy_user
-        autoconfig = self.autoconfig
-        use_sasl = configuration.conf.get('core', 'security') == 'kerberos'
-
-        try:
-            connections = self.get_connections(self.hdfs_conn_id)
-
-            if not effective_user:
-                effective_user = connections[0].login
-            if not autoconfig:
-                autoconfig = connections[0].extra_dejson.get('autoconfig',
-                                                             False)
-            hdfs_namenode_principal = connections[0].extra_dejson.get(
-                'hdfs_namenode_principal')
-        except AirflowException:
-            if not autoconfig:
-                raise
-
-        if autoconfig:
-            # will read config info from $HADOOP_HOME conf files
-            client = AutoConfigClient(effective_user=effective_user,
-                                      use_sasl=use_sasl)
-        elif len(connections) == 1:
-            client = Client(connections[0].host, connections[0].port,
-                            effective_user=effective_user, use_sasl=use_sasl,
-                            hdfs_namenode_principal=hdfs_namenode_principal)
-        elif len(connections) > 1:
-            nn = [Namenode(conn.host, conn.port) for conn in connections]
-            client = HAClient(nn, effective_user=effective_user,
-                              use_sasl=use_sasl,
-                              hdfs_namenode_principal=hdfs_namenode_principal)
-        else:
-            raise HDFSHookException("conn_id doesn't exist in the repository "
-                                    "and autoconfig is not specified")
-
-        return client
+        if self._conn is None:
+            if self.hdfs_conn_id is None:
+                self._conn = hdfs3.HDFileSystem(autoconf=self._autoconf)
+            else:
+                params = self.get_connection(self.hdfs_conn_id)
+                extra_params = params.extra_dejson
+
+                # Extract hadoop parameters from extra.
+                hdfs_params = extra_params.get("pars", {})
+
+                # Configure kerberos security if used by Airflow.
+                if configuration.conf.get("core", "security") == "kerberos":
+                    hdfs_params["hadoop.security.authentication"] = "kerberos"
+
+                # Extract high-availability config if given.
+                ha_params = extra_params.get("ha", {})
+                hdfs_params.update(ha_params.get("conf", {}))
+
+                # Collect extra parameters to pass to kwargs.
+                extra_kws = {}
+                if params.login:
+                    extra_kws["user"] = params.login
+
+                # Build connection.
+                self._conn = hdfs3.HDFileSystem(
```

Comment:
Reading through `hdfs3` docs, it seems that parameters such as `user`, `ticket_cache`, and `token` might also be useful for accessing a kerberized cluster (see [here](http://hdfs3.readthedocs.io/en/latest/_modules/hdfs3/core.html)).

These parameters, however, might need to be dag specific (i.e. a dag impersonates user_a, another user_b). So we might put them in the hook's `__init__`. What do you think?

Reply:
I'm not sure, as I would like to avoid adding too many too specific arguments. We could add a single argument `hdfs3_kwargs`, which contains kwargs that  are passed directly to `hdfs3.HDFileSystem`.

Reply:
So, if we provide them in the connection, it means we need a connection per different ticket_cache/user.

@Fokko Is this something usual in Airflow? Providing this in the dag code is more flexible.

@bolkedebruin How are you handling this at ING?